src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -615,18 +623,23 @@ private void FinishWritingHeadersUnsynchronized(int streamId, int payloadLength,
         else
         {
             // This may happen in case of the TRAILERS after the initial encode operation.
+            // The _maxFrameSize sized _headerEncodingBuffer was too small.
             while (writeResult == HPackHeaderWriter.HeaderWriteResult.BufferTooSmall)
             {
                 Debug.Assert(payloadLength == 0, "Payload written even though buffer is too small");
-                bufferSize *= 2;
-                largeHeaderBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
-                buffer = largeHeaderBuffer.AsSpan(0, bufferSize);
+                largeHeaderBuffer = ArrayPool<byte>.Shared.Rent(_headersEncodingLargeBufferSize);
+                buffer = largeHeaderBuffer.AsSpan(0, _headersEncodingLargeBufferSize);
                 writeResult = HPackHeaderWriter.RetryBeginEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out payloadLength);
                 if (writeResult != HPackHeaderWriter.HeaderWriteResult.BufferTooSmall)
                 {
                     SplitHeaderFramesToOutput(streamId, buffer[..payloadLength], endOfHeaders: writeResult == HPackHeaderWriter.HeaderWriteResult.Done, isFramePrepared: true);
                 }
+                else
+                {
+                    _headersEncodingLargeBufferSize *= HeaderBufferSizeMultiplier;
```

Comment:
Do we have a limit somewhere that will prevent this from overflowing?

Reply:
No but array allocation would throw earlier no? I will double check later today.

Reply:
With `int.MaxValue` ArrayPool throws: 'Array dimensions exceeded supported range.'
With `int.MaxValue / 2` It can still rent an array.
With negative value ArrayPool throws: 'minimumLength ('-1') must be a non-negative value. (Parameter 'minimumLength')
With value 0 it could get into an infinite loop.

So I would rather: `int.Max(_headersEncodingLargeBufferSize * HeaderBufferSizeMultiplier, _maxFrameSize);`

Updating the PR now with that.